lib/trie/trie.go

```diff
@@ -41,54 +39,64 @@ func NewTrie(root Node) *Trie {
 	}
 }
 
-// Snapshot created a copy of the trie.
-func (t *Trie) Snapshot() *Trie {
-	children := make(map[common.Hash]*Trie)
-	for h, c := range t.childTries {
-		children[h] = &Trie{
-			generation:  c.generation + 1,
-			root:        c.root,
+// Snapshot creates a copy of the trie.
+// Note it does not deep copy the trie, but will
+// copy on write as modifications are done on this new trie.
+// It does a snapshot of all child tries as well, and resets
+// the set of deleted hashes.
+func (t *Trie) Snapshot() (newTrie *Trie) {
+	childTries := make(map[common.Hash]*Trie, len(t.childTries))
+	for rootHash, childTrie := range t.childTries {
+		childTries[rootHash] = &Trie{
+			generation:  childTrie.generation + 1,
+			root:        childTrie.root.Copy(false),
 			deletedKeys: make(map[common.Hash]struct{}),
 		}
 	}
 
-	newTrie := &Trie{
+	return &Trie{
 		generation:  t.generation + 1,
 		root:        t.root,
-		childTries:  children,
+		childTries:  childTries,
 		deletedKeys: make(map[common.Hash]struct{}),
 	}
-
-	return newTrie
 }
 
-func (t *Trie) maybeUpdateGeneration(n Node) Node {
-	if n == nil {
-		return nil
```

Comment:
Since there are nil checks before for every call to `maybeUpdateGeneration`.
Before we would run `maybeUpdateGeneration` then check for nil and mostly just return nil from the caller, so this moved above; there was no point calling `maybeUpdateGeneration` when we could just check for nil and return beforehand.